File: algorithms/dynamic-programming/partition-problem/swift/CanPartition.swift
func canPartition(_ arr: [Int]) -> Int {
    let total = arr.reduce(0, +)
    if total % 2 != 0 { return 0 }
    let target = total / 2
    var dp = [Bool](repeating: false, count: target + 1)
    dp[0] = true
    for num in arr {
        for j in stride(from: target, through: num, by: -1) {
            dp[j] = dp[j] || dp[j - num]
        }
    }
    return dp[target] ? 1 : 0
}

File: algorithms/dynamic-programming/partition-problem/typescript/canPartition.ts
export function canPartition(arr: number[]): number {
    const total = arr.reduce((a, b) => a + b, 0);
    if (total % 2 !== 0) return 0;
    const target = total / 2;
    const dp = new Array(target + 1).fill(false);
    dp[0] = true;
    for (const num of arr) {
        for (let j = target; j >= num; j--) {
            dp[j] = dp[j] || dp[j - num];
        }
    }
    return dp[target] ? 1 : 0;
}
